shared_utils.py

```diff
@@ -117,7 +119,42 @@ def check_styles(soup: bs4.BeautifulSoup) -> None:
         warn('style_no_intro')
     if not get_elem_containing_text(soup, 'h1', 'references'):
         warn('style_no_refs')
-
+    # Check every numbered reference appears in the text in square brackets
+    dom = etree.HTML(str(soup))
+    ref_lis = dom.xpath("//h1[contains(text(), 'REFERENCES')]/following-sibling::ol/li | //h1[.//*[contains(text(), 'REFERENCES')]]/following-sibling::ol/li") 
+    ref_in_ref = set( range(1,len(ref_lis)+1)  ) 
+    ref_in_text_raw = re.findall('\[([\d, ]+)\]', soup.get_text( strip=True) )
+    ref_in_text = set( [ int(i) for i in re.split(r'\D+',','.join(ref_in_text_raw)) ] )
+    mismatched_ref = ref_in_ref.symmetric_difference(ref_in_text)
+    if len(mismatched_ref) > 0:
+        warn('mismatched_refs', sorted (mismatched_ref) )
+    # Check references are complete; executes anystyle in shell
+    refs = [ ' '.join( "".join(ref_li.itertext() ).split() ) for ref_li in ref_lis ] #lxml makes this ugly
+    fname = os.path.join(output_dir, "extracted_refs.txt")
+    with open(os.path.join(fname), 'w') as ofile:
+        for ref in refs:
+            ofile.write(ref + '\n')
+    subprocess.call([CONFIG['anystyle_path'], '-f' , 'json', 'parse',
+                    fname, output_dir], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+    # os.unlink(fname)
+    fname = os.path.join(output_dir,"extracted_refs.json")
+    with open(os.path.join(fname)) as ofile:
+        ref_dict_list = json.load(ofile)
+    # os.unlink(fname)
+    # TODO: put this in a config file or something
+    ref_requirements = defaultdict(lambda: ["title", "date"])
+    ref_requirements["book"] = [ "author", "title", "date", "publisher"]
+    ref_requirements["report"] = [ "author", "title", "date", "publisher"]
+    ref_requirements["chapter"] =  ["author", "title", "date", "publisher", "editor", "container-title", "pages", "location"]
+    ref_requirements["paper-conference"] = [ "author", "title", "date", "container-title", "pages"]
+    ref_requirements["article-journal"] = [ "author", "title", "date", "container-title", "pages", "volume", "issue"]
```

Comment:
Yes this is the 2nd system I've tried. The first was pure Python (refextract). It was so terrible that it just wasn't worth it; the new anystyle seems plausible but sometimes wrong.

I'll update the warning to qualify the errors

Reply:
Would you be fine with deleting the issue part from the journal requirements? I don't want to overwhelm authors with minor false positives, out of a fear that it will dilute the very important true positives like wrong styles or missing alt text that can really derail the conversion process.

Reply:
Oops should have refreshed this page :man_facepalming: 

I just ran a test on the template where I deleted the issue numbers that were really there and got the same warning. 

This makes me think that it is always saying this is missing no matter what (could be EDM style). So I agree turning it off for now makes sense. Will do another push shortly.